Refactor queue management to avoid race conditions

DavDo opened the pull request:
**Jira issue:**
https://glia.atlassian.net/browse/MOB-3698

**What was solved?**
In my previous change, I didn't consider that, in some circumstances, we're requesting relevant queues before we receive the site queues from the server. Therefore, I made everything reactive, improved readability and added an initialization check in EntryWidgetController to show an `Error` state immediately if the SDK is not initialized.

**Release notes:**

 - [ ] Feature
 - [x] Ignore
 - [ ] Release notes (Is it clear from the description here?)
 - [ ] Migration guide (If changes are needed for integrator already using the SDK - what needs to be communicated? Add underneath please)

**Additional info:**

 - [ ] Is the feature sufficiently tested? All tests fixed? Necessary unit, acceptance, snapshots added? Check that at least new public classes & methods are covered with unit tests
 - [ ] Did you add logging beneficial for troubleshooting of customer issues?
 - [ ] **Did you add new logging?** We would like the logging between platforms to be similar. Refer to [**Logging from Android SDKs** → **Things to consider for newly added logs**](https://glia.atlassian.net/wiki/spaces/ENG/pages/3568861468/Logging+from+Android+SDKs#Things-to-consider-for-newly-added-logs) in Confluence for more information.

**Screenshots:**

andrews-moc reviewed (commented):
> in some circumstances, we're requesting relevant queues before we receive the site queues from the server
Could you please describe these cases? Maybe we will need to include them to the test cases for exploratory testing.

DavDo commented:
> > in some circumstances, we're requesting relevant queues before we receive the site queues from the server
> 
> Could you please describe these cases? Maybe we will need to include them to the test cases for exploratory testing.

Sure! I mistakenly thought it was impossible to request an engagement right after initialization. So we can get IDs directly. But, when you tap the chat button without initialization, it will immediately initialize SDK and open the chat screen, which will start enqueueing with a (still) empty list.
I also somehow skipped the case when the Entry widget was requested without initializing the SDK, and I fixed this case as well.

andrews-moc commented on app/src/main/res/values/themes.xml:
```diff
@@ -11,7 +11,7 @@
     <style name="Application.Glia.Activity.Style" parent="Theme.MaterialComponents.DayNight.NoActionBar">
         <!-- example usage of compile time styling -->
-        <item name="gliaChatStyle">@style/Theme.GliaAndroidSdkWidgetsExample.Chat</item>
+        <!--        <item name="gliaChatStyle">@style/Theme.GliaAndroidSdkWidgetsExample.Chat</item>-->
```
👍

andrews-moc commented on widgetssdk/src/main/java/com/glia/widgets/core/queue/QueueRepository.kt:
```diff
@@ -32,57 +40,65 @@ internal class QueueRepositoryImpl(private val gliaCore: GliaCore, private val c
 
     private fun fetchQueues() {
-        if (siteQueues.value?.isSuccess != true) {
+        if (siteQueues.value?.isSuccess != true && gliaCore.isInitialized) {
```
I think it's logical to change the order to `(gliaCore.isInitialized && siteQueues.value?.isSuccess != true)`.

andrews-moc commented on widgetssdk/src/main/java/com/glia/widgets/core/queue/QueueRepository.kt:
```diff
@@ -32,57 +40,65 @@ internal class QueueRepositoryImpl(private val gliaCore: GliaCore, private val c
+        subscribeToQueueUpdates()
+    }
+
+    private fun getSiteQueuesError(exception: GliaException?) {
+        siteQueues.onNext(Result.failure(exception ?: RuntimeException("Fetching queues failed")))
+        Logger.e(TAG, "Fetching queues failed", exception)
+    }
```
NIT. From what I see, the names like `reportSiteQueuesError`, `onFetchingQueuesError()` might reflect the function responsibility better.

andrews-moc commented on widgetssdk/src/main/java/com/glia/widgets/core/queue/QueueRepository.kt:
```diff
@@ -32,57 +40,65 @@ internal class QueueRepositoryImpl(private val gliaCore: GliaCore, private val c
+        Logger.e(TAG, "Fetching queues failed", exception)
+    }
+
+    private fun getIds(queues: List<Queue>): List<String> = queues.map { it.id }
```
NIT. It's a tiny function, I think `queues.map { it.id }` can just be put without having a function.

andrews-moc commented:
> > > in some circumstances, we're requesting relevant queues before we receive the site queues from the server
> > 
> > 
> > Could you please describe these cases? Maybe we will need to include them to the test cases for exploratory testing.
> 
> Sure! I mistakenly thought it was impossible to request an engagement right after initialization. So we can get IDs directly. But, when you tap the chat button without initialization, it will immediately initialize SDK and open the chat screen, which will start enqueueing with a (still) empty list. I also somehow skipped the case when the Entry widget was requested without initializing the SDK, and I fixed this case as well.

Got it, thank you.

> I also somehow skipped the case when the Entry widget was requested without initializing the SDK, and I fixed this case as well.

This case has been handled correctly, I think. Entry Widget with an error state was shown before SDK is initialized. I've just re-tested. Or you mean something else?

DavDo commented:
> > Sure! I mistakenly thought it was impossible to request an engagement right after initialization. So we can get IDs directly. But, when you tap the chat button without initialization, it will immediately initialize SDK and open the chat screen, which will start enqueueing with a (still) empty list. I also somehow skipped the case when the Entry widget was requested without initializing the SDK, and I fixed this case as well.
> 
> Got it, thank you.
> 
> > I also somehow skipped the case when the Entry widget was requested without initializing the SDK, and I fixed this case as well.
> 
> This case has been handled correctly, I think. Entry Widget with an error state was shown before SDK is initialized. I've just re-tested. Or you mean something else?

Correct! That was because we were trying to fetch queues and were getting the `SDK is not initialized` error, but since I added a check for initialization before fetching queues, this logic got broken.
It was based on the `fetchQueues` error instead of handling initialization itself.

DavDo commented:
!squash
